File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.cs
public class Solution {
    public int CharacterReplacement(string s, int k) {
        int[] charCounts = new int[26];
        int right = 0;
        int left = 0;
        int maxFreq = 0;
        int maxLength = 0;

        while (right < s.Length && left < s.Length) {
            charCounts[s[right] - 'A']++;
            maxFreq = Math.Max(maxFreq, charCounts[s[right] - 'A']);

            bool isWindowValid = (right - left + 1) - maxFreq <= k;

            if (!isWindowValid) {
                charCounts[s[left] - 'A']--;
                left++;
            }
            maxLength = Math.Max(maxLength, right - left + 1);
            right++;
        }

        return maxLength;
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-5.java
class Solution {
    public int trap(int[] height) {

        if(height.length == 0){
            return 0;
        }

        int left = 0;
        int right = height.length - 1;

        int leftMax = height[left];
        int rightMax = height[right];

        int sum = 0;

        while(left < right){
            if(leftMax < rightMax){
                left++;
                leftMax = Math.max(leftMax, height[left]); 
                sum += leftMax - height[left];
            }else{
                right--;
                rightMax = Math.max(rightMax, height[right]);
                sum += rightMax - height[right];
            }
        }

        return sum;
    }
}
